src/protocol/ERC20Bridge.sol

```diff
@@ -0,0 +1,233 @@
+            decimals = _decimals;
+        } catch {
+            decimals = 18; // Standard default
+        }
+
+        TokenInitialization memory tokenInit =
+            TokenInitialization({originalToken: token, name: name, symbol: symbol, decimals: decimals});
+
+        id = _generateInitializationId(tokenInit);
+
+        // Send signal for cross-chain initialization
+        signalService.sendSignal(id);
+
+        emit TokenInitialized(id, tokenInit);
+    }
+
+    /// @inheritdoc IERC20Bridge
+    function deployCounterpartToken(TokenInitialization memory tokenInit, uint256 height, bytes memory proof)
+        external
+        returns (address deployedToken)
+    {
+        bytes32 id = _generateInitializationId(tokenInit);
+        require(!_processed[id], InitializationAlreadyProven());
+
+        // Verify the initialization signal from the source chain
+        signalService.verifySignal(height, trustedCommitmentPublisher, counterpart, id, proof);
+
+        // Mark initialization as processed
+        _processed[id] = true;
+
+        // Deploy the bridged token
+        deployedToken =
+            address(new BridgedERC20(tokenInit.name, tokenInit.symbol, tokenInit.decimals, tokenInit.originalToken));
+
+        // Store the mapping
+        bytes32 key = keccak256(abi.encode(tokenInit.originalToken));
+        _deployedTokens[key] = deployedToken;
+
+        // Mark as a bridged token deployed by this bridge
+        _isBridgedTokens[deployedToken] = true;
+
+        emit TokenInitializationProven(id, tokenInit, deployedToken);
+    }
+
+    /// @inheritdoc IERC20Bridge
+    function deposit(address to, address localToken, uint256 amount) external nonReentrant returns (bytes32 id) {
+        // Allow deposits of any token - no initialization check required
+
+        // If depositing an original token, use its address directly
+        address originalToken = localToken;
+        if (_isBridgedToken(localToken)) {
+            // If depositing a bridged token, use its original token address
+            originalToken = BridgedERC20(localToken).originalToken();
+        }
+
+        ERC20Deposit memory erc20Deposit = ERC20Deposit({
+            nonce: _globalDepositNonce,
+            from: msg.sender,
+            to: to,
+            localToken: originalToken,
+            amount: amount
+        });
+
+        id = _generateDepositId(erc20Deposit);
+        unchecked {
+            ++_globalDepositNonce;
+        }
+
+        // Handle token transfer based on whether it's a bridged token or original token
+        IERC20(localToken).safeTransferFrom(msg.sender, address(this), amount);
+        if (_isBridgedToken(localToken)) {
+            // This is a bridged token being sent back to its origin, burn it
+            BridgedERC20(localToken).burn(amount);
+        }
+
+        // Send signal
+        signalService.sendSignal(id);
+        emit DepositMade(id, erc20Deposit);
```

Comment:
this event should probably mention the `localToken` as well if it's different from the original